feat: 채용공고 CRUD

junsun-yeam opened the pull request:
- 채용공고 관련 예외 추가

## #️⃣연관된 이슈

> ex) closes #16 , closes #17 

## 📝작업 내용

> 채용공고 관련 CRUD

### 스크린샷 (선택)
<img width="2550" height="1462" alt="image" src="https://github.com/user-attachments/assets/26fcd4ed-f575-4b25-a5dd-61c7a06e20d7" />
<img width="1998" height="1391" alt="image" src="https://github.com/user-attachments/assets/2cf5cc48-6e53-427d-8656-e4af197ce892" />
<img width="2598" height="1408" alt="image" src="https://github.com/user-attachments/assets/ca4a9689-4902-46e9-8d63-e606938eec01" />


## 💬리뷰 요구사항(선택)
- 편하게 피드백 반영받겠습니다

junsun-yeam committed:
feat: 채용공고 CRUD

- 채용공고 관련 예외 추가

lesw1216 commented on src/main/java/com/halo/core_bridge/api/jobposting/controller/JobPostingController.java:
```diff
@@ -0,0 +1,37 @@
+    ) {
+       return BaseResponse.success(jobPostingService.save(request));
+    }
+
+    //채용공고 전체 목록 조회
+    @GetMapping
+    public BaseResponse<List<JobPostingDto.ListResponse>> getAllJobPostings() {
+        return BaseResponse.success(jobPostingService.getList());
+    }
```
Reponse 응답을 내려줄 때 ResponseEntity로 한번 더 감싸 주는 방법이 더 좋을 것 같아요!
지금 구조에서는 모든 응답이 기본적으로 `200 OK`로 내려가게 되는데,
이렇게 되면 REST API 설계 원칙상 성공/실패 구분이 명확하지 않아 **좋은 API 설계**라고 보기 어려워요.

`ResponseEntity`를 사용하면 `ok()`, `created()`, `badRequest()` 등 다양한 상태 코드를 지정할 수 있어서 더 RESTful한 방식으로 응답을 내려줄 수 있어요!

`ResponseEnity`나 REST API 응답 구조 등 관련해서 한번 찾아보시면 도움이 많이 될 것 같아요!

```suggestion
    @GetMapping
    public ResponseEntity<BaseResponse<List<JobPostingDto.ListResponse>>> getAllJobPostings() {
    
        return ResponseEntity.ok(BaseResponse.success(jobPostingService.getList()));
    }
```

lesw1216 commented on src/main/java/com/halo/core_bridge/api/jobposting/service/JobPostingService.java:
```diff
@@ -0,0 +1,47 @@
+                .orElseThrow(() -> BaseException.from(BaseResponseStatus.DEPARTMENT_NOT_FOUND));
+
+        var entity = dto.toEntity(dept);
+        var saved = jobPostingRepository.save(entity);
```
var 대신 명시적인 타입을 작성해주시면 더 명확할 것 같아요.

현재 두 메서드가 어떤 타입을 반환하는지 코드만 봐서는 바로 파악하기 어려워요
명시적인 타입을 작성하면 가독성이 좋아지고, IDE 도움 없이 코드를 읽을 때도 의도를 쉽게 이해할 수 있어요.

```java
JobPosting JobPostingEntity= dto.toEntity(dept);
JobPosting savedJobPosting = jobPostingRepository.save(JobPostingEntity);
```

또한 변수명도 entity, saved보다
JobPostingEntity, savedJobPosting처럼 의미 있는 이름을 사용하면
이후 유지보수 시 코드 맥락을 훨씬 빠르게 파악할 수 있을 것 같아요.

lesw1216 commented on src/main/java/com/halo/core_bridge/api/jobposting/service/JobPostingService.java:
```diff
@@ -0,0 +1,47 @@
+    // 채용공고 등록
+    public JobPostingDto.DetailResponse save(JobPostingDto.CreateRequest dto) {
+        Department dept = departmentRepository.findById(dto.getDepartmentId())
+                .orElseThrow(() -> BaseException.from(BaseResponseStatus.DEPARTMENT_NOT_FOUND));
```
`JobPosting`을 저장할 때 `Department` 정보를 직접 조회해서 넣으신 것 같아요!

JPA에서는 연관관계가 설정되어 있다면, **연관된 엔티티의 식별자(id)** 만으로도 매핑이 가능해요.
그래서 `Department` 엔티티 전체를 조회하지 않고, **id 값만 세팅해도** `JobPosting` 저장 시 정상적으로 연관관계가 매핑돼요.

이렇게 되면 불필요한 `Department` 조회 쿼리를 줄일 수 있어요! 
이부분에 관해서는 **JPA의 getReference()** 부분을 찾아보시면 도움이 되실 것 같아요!

그리고 **설계적인 관점**에서는,
하나의 서비스 클래스에서 다른 엔티티의 `Repository`를 직접 사용하는 대신
해당 엔티티 전용 `Service`를 만들어 위임하는 방식도 고려해보면 좋을 것 같아요.

이렇게 분리하면 `JobPostingService`가 `Department` 관련 로직까지 책임지지 않게 되어
역할이 명확해지고 유지보수성도 높아져요.

lesw1216 commented on src/main/java/com/halo/core_bridge/api/organization/repository/DepartmentRepository.java:
현재 프로젝트 구조를 보면 Organization 패키지 안에 부서 관련 엔티티들이 모여 있는 것 같아요

이 파일도 해당 패키지 내부에 위치시키면 도메인 단위로 코드가 정리되어 구조적으로 더 일관성 있게 유지될 것 같아요!

패키지 구조가 명확해지면, 이후에 관련 기능을 찾거나 관리할 때도 훨씬 편리할 것 같습니다!

lesw1216 commented on src/main/java/com/halo/core_bridge/api/jobposting/model/dto/JobPostingDto.java:
```diff
@@ -0,0 +1,104 @@
+
+    // 공고 목록 응답
+    @Getter
+    @NoArgsConstructor @AllArgsConstructor @Builder
```
현재 클래스에 Lombok 어노테이션이
@Getter, @NoArgsConstructor, @AllArgsConstructor, @Builder
이렇게 한 줄에 모여 있는데,
각 클래스마다 어노테이션 일관성을 가져가면서 각 어노테이션의 역할을 명확히 보기 위해 한 줄씩 나누어 작성하면 좋을 것 같아요!
```suggestion
    @Getter
    @NoArgsConstructor 
    @AllArgsConstructor 
    @Builder
```

lesw1216 commented on src/main/java/com/halo/core_bridge/api/jobposting/model/entity/JobPosting.java:
엔티티의 필드에 컬럼 속성을 명시해주시는 건 어떨까요?

`@Column`을 사용해서 `nullable`, `length`, `unique` 등의 속성을 지정하면
DB 스키마 설계 의도가 코드 상에서도 명확하게 드러나요!

예를 들어

```java
@Column(nullable = true)
private Integer minExperience;

@Column(nullable = true)
private Integer maxExperience;
```

이런 식으로 지정해두면,
JPA가 스키마 자동 생성 시 정확한 제약 조건을 적용하고,
다른 개발자도 해당 컬럼이 선택적 값임을 바로 인지할 수 있어요!

lesw1216 commented on src/main/java/com/halo/core_bridge/api/jobposting/model/entity/JobPosting.java:
```diff
@@ -19,16 +18,29 @@ public class JobPosting extends BaseEntity {
+
     private LocalDateTime applyStartDate;
     private LocalDateTime applyEndDate;
     private LocalDateTime hireEndDate;
+
     private int minExperience;
     private int maxExperience;
```
`minExperience`, `maxExperience` 컬럼은 채용공고에서 경력 조건(최소·최대 년차) 을 나타내는 컬럼으로 지정했던 기억이 있어요!

신입이나 경력 무관 공고의 경우 이 값이 없을 수 있기 때문에,
해당 컬럼은 `nullable = true` 로 설정해주시면 더 유연할 것 같아요!

이렇게 하면 선택적으로만 값이 들어가고,
불필요한 0년차 값 저장을 피할 수 있을 것 같아요.

lesw1216 commented on src/main/java/com/halo/core_bridge/api/jobposting/model/entity/JobPosting.java:
```diff
@@ -19,16 +18,29 @@ public class JobPosting extends BaseEntity {
     private int minExperience;
     private int maxExperience;
+
     private String employmentType;
     private String careerType;
```
이 부분은 제가 깜빡하고 `Enum` 타입을 생성하지 않았나봐요 ㅠ
제가 Enum 생성해서 고쳐둘게요!

lesw1216 commented on src/main/java/com/halo/core_bridge/api/jobposting/service/JobPostingService.java:
```diff
@@ -0,0 +1,47 @@
+
+    // 전체 목록 조회
+    public List<JobPostingDto.ListResponse> getList() {
+        List<JobPosting> list = jobPostingRepository.findAll();
```
findAll()의 반환값을 list라는 변수명으로 사용하셨네요.

변수명을 list로만 지정하면, 컬렉션 타입(List) 과 이름이 겹쳐서
코드를 읽을 때 혼동될 수 있어요

예를 들어 아래처럼 코드만 보면
“list가 타입명인지 변수명인지” 직관적으로 구분하기 어려워요.

```java
List<JobPosting> list = jobPostingRepository.findAll();
```

조금 더 구체적이고 의미 있는 이름을 사용하면 가독성이 크게 좋아집니다 💡

```java
List<JobPosting> jobPostings = jobPostingRepository.findAll();
```

이렇게 하면 코드만 봐도 어떤 데이터가 담겨 있는지 명확해지고,
이후 로직에서 반복문이나 스트림을 사용할 때도 의도가 더 잘 드러나요!

lesw1216 reviewed (changes requested):
요청을 받는 DTO에는 **입력값 검증(Validation)** 어노테이션을 추가해주시면 좋을 것 같아요.

서버 단에서 한 번 유효성 검사를 거치면
클라이언트에서 잘못된 데이터가 들어오는 걸 미리 차단할 수 있고,
**보안성과 데이터 무결성 측면에서도 훨씬 안전해집니다.**

예를 들어 아래처럼 `@NotBlank`, `@Size`, `@PositiveOrZero` 등을 활용할 수 있어요.

```java
@NotBlank(message = "제목은 필수 입력값입니다.")
private String title;

@PositiveOrZero(message = "최소 경력은 0 이상이어야 합니다.")
private int minExperience;
```

그리고 응답을 내려줄 때 날짜 포맷이 현재 `YYYY-MM-dd'T'HH:mm:ss` 형태로 내려오고 있는데,
이 형식은 프론트엔드에서 가공하기 번거로울 수 있어요!
**서버 쪽에서 미리 포맷을 맞춰서 내려주면 더 깔끔할 것 같아요!**

`@JsonFormat` 또는 커스텀 `@JsonSerializer` 관련된 부분 찾아보시면 도움이 되실것 같아요!

```java
@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
private LocalDateTime applyEndDate;
```

공통적으로 날짜 형태를 일관되게 내려주기 위해
`CustomDateSerializer` 클래스를 만들어두고 `@JsonSerialize`로 연결하는 방법도 좋을 것 같아요!

```java
@JsonSerialize(using = CustomDateSerializer.class)
private LocalDateTime openDate;
```

이렇게 설정해두면 프론트와 서버 간의 날짜 포맷 일관성을 유지하면서
DTO마다 반복되는 포맷 지정도 줄일 수 있을 것 같아요!

lesw1216 commented on src/main/java/com/halo/core_bridge/api/jobposting/model/dto/JobPostingDto.java:
```diff
@@ -0,0 +1,104 @@
+                    .employmentType(this.employmentType)
+                    .careerType(this.careerType)
+                    .minExperience(this.minExperience)
+                    .maxExperience(this.maxExperience)
+                    .applyStartDate(this.applyStartDate)
+                    .applyEndDate(this.applyEndDate)
+                    .hireEndDate(this.hireEndDate)
+                    .department(department)
+                    .build();
+
+            if (skills != null) {
+                this.skills.forEach(skillName ->
+                        posting.addSkill(JobPostingSkill.builder().name(skillName).build()));
+            }
+            return posting;
+        }
+
+    }
```
`Department`는 이전 리뷰에서 말씀드린 것처럼,
**연관관계 매핑 시 Id 값만으로도 참조가 가능하기 때문에**
**파라미터로 `Department` 객체를 직접 받지 않고**
DTO에 포함된 departmentId를 이용해 **내부에서 엔티티를 생성**해주시면 좋을 것 같아요.

이렇게 하면 불필요한 `findById()` 쿼리를 줄일 수 있고,
`toEntity()` 메서드 자체가 더 독립적으로 동작할 수 있어요.

```suggestion
        public JobPosting toEntity(Department department) {
            JobPosting posting = JobPosting.builder()
                    . . .
                    .department(
                      // Department 빌더 패턴을 사용해서 DepartmentId만 담긴 Department 엔티티를 저장하면 좋을 것 같아요!
                    )
                    .build();

            . . .

    }
```

lesw1216 reviewed (changes requested):
`JobPosting` 저장 시 `skills`를 함께 생성하고 저장하는 구조로 작성해주신 것 같아요.

현재는 `cascade = ALL` 덕분에 정상적으로 동작하긴 하지만,
JPA가 엔티티 그래프 전체를 한 번에 `persist()`하면서
**빌더나 연관관계 세팅이 누락되면 FK가 깨질 위험이 있어요.**

특히 `JobPostingSkill`의 `setJobPosting()`이나
`JobPosting`의 `addSkill()`처럼 외부에서 연관관계를 직접 수정할 수 있는 구조는
**엔티티 일관성을 깨뜨릴 가능성**이 있어서 setter 성향의 메서드는 최소화하는 게 좋을 것 같아요.

또한 `JobPosting`의 `skills` 필드는 `mappedBy`로 지정되어 연관관계 주인이 아니기 때문에 실제 저장시에는 무시되는 필드에요!
따라서 **단순 조회용(read-only)** 역할로 사용하는 편이 좋을 것 같아요!

저장 로직에서는 다음과 같은 흐름으로 책임을 분리해보시면 좋을 것 같아요!

1. 채용공고(JobPosting)를 먼저 저장
2. 저장된 공고의 id를 이용하여 요청 DTO(`JobPostingDto.CreateRequest`) 내의 `skills`를 `JobPostingSkillService`에서 저장

이런 구조로 변경하면

* 엔티티 내부에서 setter 성향의 코드를 제거할 수 있고
* `JobPosting`은 순수한 공고 정보만 관리하게 되며
* `JobPosting`의 `skills`는 읽기 전용으로 명확히 역할이 분리되요

결과적으로 엔티티의 책임이 명확해지고,
`skills`는 읽기 전용 참조로만 사용되기 때문에
JPA 연관관계의 의도에 더 맞는 구조가 될 것 같아요!

junsun-yeam committed:
fix: CRUD 수정

- 요청 dto에 대한 입력값 검증
- ResponseEntity로 응답
- 명시적 타입 -> 구체적 타입으로 변경
- department Service로직 분리
- 엔티티 필드에 컬럼 속성 명시
- 명확한 변수명으로 변경
- 채용공고 저장 후 jobpostingskill 저장

junsun-yeam committed:
fix: CRUD 수정

- 요청 dto에 대한 입력값 검증
- ResponseEntity로 응답
- 명시적 타입 -> 구체적 타입으로 변경
- 날짜형태(yyyy-MM-dd HH:mm:ss)로 통일
- department Service로직 분리
- 엔티티 필드에 컬럼 속성 명시
- 명확한 변수명으로 변경
- 채용공고 저장 후 jobpostingskill 저장
